File: tests/cpp/test_resolution_plan.cpp
#include <cstdio>
#include <cstdlib>
#include <string>

#include "resolution/resolution_plan.h"

namespace
{

void require(bool condition, const char* message)
{
    if (!condition)
    {
        std::fprintf(stderr, "FAIL: %s\n", message);
        std::exit(1);
    }
}

} // namespace

int main()
{
    seedvr2::ResolutionPlan square;
    require(seedvr2::ResolutionPlan::from_explicit(128, 128, square), "explicit square plan");
    require(square.image_height == 128 && square.image_width == 128, "square image shape");
    require(square.latent_height == 16 && square.latent_width == 16, "square latent shape");
    require(square.source_t == 1 && square.source_height == 8 && square.source_width == 8,
            "square dit shape");
    require(square.video_tokens == 64, "square token count");

    seedvr2::ResolutionPlan widescreen;
    require(seedvr2::ResolutionPlan::from_explicit(720, 1280, widescreen), "explicit widescreen plan");
    require(widescreen.latent_height == 90 && widescreen.latent_width == 160, "widescreen latent shape");
    require(widescreen.source_t == 1 && widescreen.source_height == 45 && widescreen.source_width == 80,
            "widescreen dit shape");
    require(widescreen.video_tokens == 3600, "widescreen token count");

    seedvr2::ResolutionPlan automatic;
    require(seedvr2::ResolutionPlan::from_input_area(720, 1280, automatic), "official auto plan");
    require(automatic.image_height == 720 && automatic.image_width == 1280, "official auto dimensions");
    require(automatic.crop_top == 0 && automatic.crop_left == 0, "official auto crop");

    seedvr2::ResolutionPlan upsampled;
    require(seedvr2::ResolutionPlan::from_input_area(100, 100, upsampled), "small auto plan");
    require(upsampled.image_height == 960 && upsampled.image_width == 960, "small image area resize");
    require(upsampled.latent_height == 120 && upsampled.latent_width == 120, "small image latent shape");

    seedvr2::ResolutionPlan invalid;
    require(!seedvr2::ResolutionPlan::from_explicit(0, 128, invalid), "zero height rejected");
    require(!seedvr2::ResolutionPlan::from_explicit(128, 127, invalid), "unaligned width rejected");
    require(!seedvr2::ResolutionPlan::from_input_area(-1, 128, invalid), "negative input rejected");

    std::puts("seedvr2-resolution: ok");
    return 0;
}

File: tests/python/test_resolution_plan.py
import pytest

from tools.reference.resolution_plan import plan_from_explicit, plan_from_input_area, round_half_even


def test_explicit_plan_derives_latent_and_dit_grid():
    plan = plan_from_explicit(720, 1280)

    assert (plan.image_height, plan.image_width) == (720, 1280)
    assert (plan.latent_height, plan.latent_width) == (90, 160)
    assert (plan.source_shape, plan.video_tokens) == ((1, 45, 80), 3600)


def test_auto_plan_matches_seedvr2_area_resize_and_divisible_crop():
    plan = plan_from_input_area(720, 1280)

    assert (plan.resized_height, plan.resized_width) == (720, 1280)
    assert (plan.image_height, plan.image_width) == (720, 1280)
    assert (plan.crop_top, plan.crop_left) == (0, 0)


def test_auto_plan_upsamples_small_image_before_alignment():
    plan = plan_from_input_area(100, 100)

    assert (plan.resized_height, plan.resized_width) == (960, 960)
    assert (plan.image_height, plan.image_width) == (960, 960)
    assert (plan.latent_height, plan.latent_width) == (120, 120)


@pytest.mark.parametrize("height,width", [(0, 128), (-1, 128), (128, 127), (736, 1280)])
def test_invalid_dimensions_are_rejected(height, width):
    with pytest.raises(ValueError):
        plan_from_explicit(height, width)


def test_auto_plan_uses_half_even_rounding_for_resize():
    assert round_half_even(2.5) == 2
    assert round_half_even(3.5) == 4
